dace/transformation/transformation.py

```diff
@@ -351,81 +351,204 @@ def print_match(self, sdfg: SDFG) -> str:
+        """
+        pass
+
+
+@make_properties
+class MultiStateTransformation(PatternTransformation, abc.ABC):
+    """
+    Base class for pattern-matching transformations that find matches within an SDFG state machine.
+    New transformations that extend this class must contain static ``PatternNode``-annotated fields that represent the
+    nodes in the pattern graph, and use them to implement at least three methods:
+        * ``expressions``: A method that returns a list of graph
+                           patterns (SDFG objects) that match this
+                           transformation.
+        * ``can_be_applied``: A method that, given a subgraph candidate,
+                              checks for additional conditions whether it can
+                              be transformed.
+        * ``apply``: A method that applies the transformation
+                     on the given SDFG.
+
+    For example:
+    ```
+    class MyTransformation(MultiStateTransformation):
+        state_a = PatternNode(SDFGState)
+        state_b = PatternNode(SDFGState)
+
+        @classmethod
+        def expressions(cls):
+            return [node_path_graph(cls.state_a, cls.state_b)]
+    ```
+
+    For more information and optimization opportunities, see the respective
+    methods' documentation.
+
+    :seealso: PatternNode
+    :note: Some transformations are included in the SDFG Dataflow Coarsening pass. In order to declare a 
+           transformation as part of the dataflow coarsening pass, it should also extend the 
+           ``DataflowCoarseningTransformation`` mixin.
+    """
+    @classmethod
+    @abc.abstractmethod
+    def expressions(cls) -> List[gr.SubgraphView]:
+        """ Returns a list of SDFG subgraphs that will be matched in the
+            subgraph isomorphism phase. Used as a pre-pass before calling
+            ``can_be_applied``.
+        """
+        pass
+
+    @abc.abstractmethod
+    def can_be_applied(self, graph: SDFG, expr_index: int, sdfg: SDFG, permissive: bool = False) -> bool:
+        """ Returns True if this transformation can be applied on the candidate matched subgraph.
+            :param graph: SDFG object in which the match was found.
+            :param candidate: A mapping between node IDs returned from
+                              ``PatternTransformation.expressions`` and the nodes in
+                              ``graph``.
+            :param expr_index: The list index from ``PatternTransformation.expressions``
+                               that was matched.
+            :param sdfg: The SDFG in which the match was found (equal to ``graph``).
+            :param permissive: Whether transformation should run in permissive mode.
+            :return: True if the transformation can be applied.
+        """
+        pass
+
+
+class DataflowCoarseningTransformation:
+    """
+    Mixin that includes the given PatternTransformation in the SDFG Dataflow Coarsening pass. This transformation
```

Comment:
Isn't a "mixin" basically an interface? This class has no methods, therefore, it is not an interface. Wouldn't it be better to just make it a boolean class attribute (in PatternTransformation)?

Reply:
Yes, but the plan is to perhaps make more such Passes that transformations can participate in. In the new PR it's renamed to be `SimplifyPass`